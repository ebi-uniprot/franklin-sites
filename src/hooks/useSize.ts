// NOTE: same as in uniprot-website codebase
import { MutableRefObject, useCallback, useEffect, useState } from 'react';

/**
 * given a reference to an actual HTML element, gives us access to its size
 * even through resizes and zoom
 * @param {MutableRefObject<HTMLElement>} ref
 * @returns {Array} tuple - return tuple
 * @returns {DOMRect | null} tuple[0]
 *          DOMRect object with the sizes of the passed ref, or null if no ref
 * @returns {() => void} tuple[1]
 *          Manual refresh function, to call if the size needs to be recomputed
 *          for whatever reason (e.g. content changed)
 */
function useSize(
  ref: MutableRefObject<HTMLElement | null>
): [rect: DOMRect | undefined, onResize: () => void] {
  const [rect, setRect] = useState<DOMRect>();

  const onResize = useCallback(() => {
    if (!ref.current) {
      return;
    }
    setRect(ref.current.getBoundingClientRect());
  }, [ref, setRect]);

  // This effect will run after EVERY render
  useEffect(() => {
    // ask for size value until we finally get one
    if (!rect) {
      onResize();
    }
  }); // No dependency array, it's on purpose!

  useEffect(() => {
    onResize(); // first time
    window.addEventListener('resize', onResize);
    // eslint-disable-next-line consistent-return
    return () => window.removeEventListener('resize', onResize);
  }, [ref, onResize]);

  return [rect, onResize];
}

export default useSize;
