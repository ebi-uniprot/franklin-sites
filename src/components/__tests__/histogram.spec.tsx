import { render, screen } from '@testing-library/react';

import { Histogram } from '..';

import useSize from '../../hooks/useSize';

jest.mock('../../hooks/useSize', () => jest.fn());
(useSize as jest.Mock).mockImplementation(() => [{ width: 250, height: 100 }]);

describe('Histogram component', () => {
  it('should render when nBins prop is provided and with the correct number of bins', () => {
    const { asFragment } = render(
      <Histogram
        values={[...Array(10).keys()]}
        nBins={2}
        xLabel="x axis"
        yLabel="y axis"
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getAllByTestId('histogram-bar')).toHaveLength(2);
    expect(screen.getByText('x axis')).toBeInTheDocument();
    expect(screen.getByText('y axis')).toBeInTheDocument();
  });

  it('should render without axes if no label provided', () => {
    const { asFragment } = render(
      <Histogram values={[...Array(10).keys()]} nBins={2} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getAllByTestId('histogram-bar')).toHaveLength(2);
    expect(screen.queryByText('x axis')).not.toBeInTheDocument();
    expect(screen.queryByText('y axis')).not.toBeInTheDocument();
  });

  it('should render when binSize prop is provided and with the correct number of bins', () => {
    const { asFragment } = render(
      <Histogram
        values={[...Array(10).keys()]}
        binSize={2}
        xLabel="x axis"
        yLabel="y axis"
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getAllByTestId('histogram-bar')).toHaveLength(5);
  });

  it('should shade bins that are in range', () => {
    render(
      <Histogram
        values={[...Array(10).keys()]}
        binSize={2}
        selectedRange={[3, 7]}
        xLabel="x axis"
        yLabel="y axis"
      />
    );
    const bars = screen.getAllByTestId('histogram-bar');
    expect(
      bars.filter((bar) =>
        bar.classList.contains('histogram__bar--within-range')
      )
    ).toHaveLength(3);
  });
});
