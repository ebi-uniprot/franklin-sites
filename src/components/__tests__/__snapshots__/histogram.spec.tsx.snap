// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Histogram component should render when binSize prop is provided and with the correct number of bins 1`] = `
<DocumentFragment>
  <div
    class="histogram"
    data-testid="histogram"
  >
    <div
      class="histogram__bar-container"
      style="height: 300px;"
    >
      <div
        class="histogram histogram__bar histogram__bar--within-range"
        data-testid="histogram-bar"
        style="transform: scaleY(0.4); transition-delay: 0ms;"
      />
      <div
        class="histogram histogram__bar histogram__bar--within-range"
        data-testid="histogram-bar"
        style="transform: scaleY(0.4); transition-delay: 100ms;"
      />
      <div
        class="histogram histogram__bar histogram__bar--within-range"
        data-testid="histogram-bar"
        style="transform: scaleY(0.4); transition-delay: 200ms;"
      />
      <div
        class="histogram histogram__bar histogram__bar--within-range"
        data-testid="histogram-bar"
        style="transform: scaleY(0.4); transition-delay: 300ms;"
      />
      <div
        class="histogram histogram__bar histogram__bar--within-range"
        data-testid="histogram-bar"
        style="transform: scaleY(0.4); transition-delay: 400ms;"
      />
    </div>
    <svg
      height="80"
      style="top: 300px;"
      width="100%"
    >
      <g
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="middle"
      >
        <path
          class="domain"
          d="M0.5,0.5H250.5"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="1"
          transform="translate(0.5,0)"
        >
          <line
            stroke="currentColor"
            y2="0"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="6"
          >
            0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(56.05555555555555,0)"
        >
          <line
            stroke="currentColor"
            y2="0"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="6"
          >
            2
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(111.6111111111111,0)"
        >
          <line
            stroke="currentColor"
            y2="0"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="6"
          >
            4
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(167.16666666666666,0)"
        >
          <line
            stroke="currentColor"
            y2="0"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="6"
          >
            6
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(222.7222222222222,0)"
        >
          <line
            stroke="currentColor"
            y2="0"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="6"
          >
            8
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(278.27777777777777,0)"
        >
          <line
            stroke="currentColor"
            y2="0"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="6"
          >
            10
          </text>
        </g>
      </g>
      <text
        style="text-anchor: middle;"
        transform="translate(125, 40)"
      >
        x axis
      </text>
    </svg>
    <svg
      class="y-axis"
      height="300"
      width="80"
    >
      <g
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="end"
        transform="translate(50, 0)"
      >
        <path
          class="domain"
          d="M0.5,300.5V0.5"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="1"
          transform="translate(0,300.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            0.0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,270.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            0.5
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,240.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            1.0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,210.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            1.5
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,180.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            2.0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,150.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            2.5
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,120.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            3.0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,90.50000000000001)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            3.5
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,60.499999999999986)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            4.0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,30.499999999999993)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            4.5
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,0.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            5.0
          </text>
        </g>
      </g>
      <text
        style="text-anchor: middle;"
        transform="rotate(-90)"
        x="-150"
        y="20"
      >
        y axis
      </text>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Histogram component should render when nBins prop is provided and with the correct number of bins 1`] = `
<DocumentFragment>
  <div
    class="histogram"
    data-testid="histogram"
  >
    <div
      class="histogram__bar-container"
      style="height: 300px;"
    >
      <div
        class="histogram histogram__bar histogram__bar--within-range"
        data-testid="histogram-bar"
        style="transform: scaleY(1); transition-delay: 0ms;"
      />
      <div
        class="histogram histogram__bar histogram__bar--within-range"
        data-testid="histogram-bar"
        style="transform: scaleY(1); transition-delay: 250ms;"
      />
    </div>
    <svg
      height="80"
      style="top: 300px;"
      width="100%"
    >
      <g
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="middle"
      >
        <path
          class="domain"
          d="M0.5,0.5H250.5"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="1"
          transform="translate(0.5,0)"
        >
          <line
            stroke="currentColor"
            y2="0"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="6"
          >
            0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(125.5,0)"
        >
          <line
            stroke="currentColor"
            y2="0"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="6"
          >
            5
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(250.5,0)"
        >
          <line
            stroke="currentColor"
            y2="0"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="6"
          >
            9
          </text>
        </g>
      </g>
      <text
        style="text-anchor: middle;"
        transform="translate(125, 40)"
      >
        x axis
      </text>
    </svg>
    <svg
      class="y-axis"
      height="300"
      width="80"
    >
      <g
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="end"
        transform="translate(50, 0)"
      >
        <path
          class="domain"
          d="M0.5,300.5V0.5"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="1"
          transform="translate(0,300.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            0.0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,270.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            0.5
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,240.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            1.0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,210.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            1.5
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,180.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            2.0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,150.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            2.5
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,120.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            3.0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,90.50000000000001)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            3.5
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,60.499999999999986)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            4.0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,30.499999999999993)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            4.5
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,0.5)"
        >
          <line
            stroke="currentColor"
            x2="0"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-6"
          >
            5.0
          </text>
        </g>
      </g>
      <text
        style="text-anchor: middle;"
        transform="rotate(-90)"
        x="-150"
        y="20"
      >
        y axis
      </text>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Histogram component should render without axes if no label provided 1`] = `
<DocumentFragment>
  <div
    class="histogram"
    data-testid="histogram"
  >
    <div
      class="histogram__bar-container"
      style="height: 300px;"
    >
      <div
        class="histogram histogram__bar histogram__bar--within-range"
        data-testid="histogram-bar"
        style="transform: scaleY(1); transition-delay: 0ms;"
      />
      <div
        class="histogram histogram__bar histogram__bar--within-range"
        data-testid="histogram-bar"
        style="transform: scaleY(1); transition-delay: 250ms;"
      />
    </div>
  </div>
</DocumentFragment>
`;
